keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2397 @@
+again and provide the additional information.
+
+Additional heuristics are possible without changing the proposed API. For
+example, the scheduler might ask for information and wait a while before
+making a choice. This may be more suitable for pods using many different
+resource claims because for those, allocation may succeed for some claims and
+fail for others, which then may need to go through the recovery flow with
+deallocating one or more claims.
+
+### Resource allocation and usage flow
+
+The following steps shows how resource allocation works for a resource that
+gets defined inline in a Pod. Several of these steps may fail without changing
+the system state. They then must be retried until they succeed or something
+else changes in the system, like for example deleting objects.
+
+<<[UNRESOLVED pohly]>>
+Should the entire document use JSON field names instead of Go field names and structs?
+https://github.com/kubernetes/enhancements/pull/3064#discussion_r901950774
+<<[/UNRESOLVED]>>
+
+
+* **user** creates Pod with inline ResourceClaimTemplate
+* **resource claim controller** checks ResourceClaimTemplate and ResourceClass,
+  then creates ResourceClaim with Pod as owner
+* if *immediate allocation*:
+  * **resource driver** adds finalizer to claim to prevent deletion -> allocation in progress
+  * **resource driver** finishes allocation, sets `claim.status.allocation` -> claim ready for use by any pod
+* if *pod is pending*:
+  * **scheduler** filters nodes
+  * if *delayed allocation and resource not allocated yet*:
+    * if *at least one node fits pod*:
+      * **scheduler** creates or updates a `PodScheduling` object with `potentialNodes=<all nodes that fit the pod>`
+      * **scheduler** picks one node, sets `claim.status.scheduling.selectedNode=<the chosen node>` and `claim.status.scheduling.selectedUser=<the pod being scheduled>`
+      * if *resource is available for `claim.status.scheduling.selectedNode`*:s
+        * **resource driver** adds finalizer to claim to prevent deletion -> allocation in progress
+        * **resource driver** finishes allocation, sets `claim.status.allocation` and the
+          intended user in `claim.status.reservedFor`, clears `claim.status.selectedNode` and `claim.status.selectedUser` -> claim ready for use and reserved
+          for the pod
+      * else *scheduler needs to know that it must avoid this and possibly other nodes*:
+        * **resource driver** retrieves `PodScheduling` object from informer cache or API server
+        * **resource driver** sets `podScheduling.claims[name=name of claim in pod].unsuitableNodes`
+        * **resource driver** clears `claim.status.selectedNode` -> next attempt by scheduler has more information and is more likely to succeed
+    * else *pod cannot be scheduled*:
+      * **scheduler** may trigger deallocation of some claim with delayed allocation by setting `claim.status.deallocate` to true
```

Comment:
I will have to stare at this even harder. I think I get what you are doing, but I feel like there has to be a way to do all of the coordination in the coordination object without triggering any actual allocations until we have a solution.

Reply:
For others: we are discussing this in https://github.com/pohly/enhancements/pull/13